src/java.base/share/classes/java/util/stream/Gatherer.java

```diff
@@ -0,0 +1,593 @@
+     * @return the new {@code Gatherer}
+     */
+    static <T, R> Gatherer<T, Void, R> of(Integrator<Void, T, R> integrator) {
+        return of(
+                defaultInitializer(),
+                integrator,
+                Gatherers.Value.DEFAULT.statelessCombiner,
+                defaultFinisher()
+        );
+    }
+
+    /**
+     * Returns a new, parallelizable, and stateless {@code Gatherer} described
+     * by the given {@code integrator} and {@code finisher}.
+     *
+     * @param integrator the integrator function for the new gatherer
+     * @param finisher the finisher function for the new gatherer
+     * @param <T> the type of input elements for the new gatherer
+     * @param <R> the type of results for the new gatherer
+     * @throws NullPointerException if any argument is {@code null}
+     * @return the new {@code Gatherer}
+     */
+    static <T, R> Gatherer<T, Void, R> of(
+            Integrator<Void, T, R> integrator,
+            BiConsumer<Void, Downstream<? super R>> finisher) {
+        return of(
+                defaultInitializer(),
+                integrator,
+                Gatherers.Value.DEFAULT.statelessCombiner,
+                finisher
+        );
+    }
+
+    /**
+     * Returns a new, parallelizable, {@code Gatherer} described by the given
+     * {@code initializer}, {@code integrator}, {@code combiner} and
+     * {@code finisher}.
+     *
+     * @param initializer the initializer function for the new gatherer
+     * @param integrator the integrator function for the new gatherer
+     * @param combiner the combiner function for the new gatherer
+     * @param finisher the finisher function for the new gatherer
+     * @param <T> the type of input elements for the new gatherer
+     * @param <A> the type of state for the new gatherer
+     * @param <R> the type of results for the new gatherer
+     * @throws NullPointerException if any argument is {@code null}
+     * @return the new {@code Gatherer}
+     */
+    static <T, A, R> Gatherer<T, A, R> of(
+            Supplier<A> initializer,
+            Integrator<A, T, R> integrator,
+            BinaryOperator<A> combiner,
+            BiConsumer<A, Downstream<? super R>> finisher) {
+        return new Gatherers.GathererImpl<>(
+                Objects.requireNonNull(initializer),
+                Objects.requireNonNull(integrator),
+                Objects.requireNonNull(combiner),
+                Objects.requireNonNull(finisher)
+        );
+    }
+
+    /**
+     * A Downstream object is the next stage in a pipeline of operations,
+     * to which elements can be sent.
+     * @param <T> the type of elements this downstream accepts
+     * @since 22
+     */
+    @FunctionalInterface
+    @PreviewFeature(feature = PreviewFeature.Feature.STREAM_GATHERERS)
+    interface Downstream<T> {
+
+        /**
+         * Pushes, if possible, the provided element downstream -- to the next
```

Comment:
can it be clarified that this method can be called on the drainer thread only. Given `mapConcurrent` [implementation](https://github.com/openjdk/jdk/pull/16420/files#diff-ff65ac4f94ce5409d99530fb8b7526349846499d6fb6f6727f5668f708e6bad1R378) it seems to be like that otherwise I wounder why there is not option for unorder mapConcurrent where the `createTaskFor` method is implemented as follows :

```java
final void createTaskFor(T element, Downstream<R> downstream) {
                 windowLock.acquireUninterruptibly();

                 var task = new FutureTask<R>(() -> {
                     try {
                         downstream.push(mapper.apply(element));
                         return null 
                     } finally {
                         windowLock.release();
                     }
                 });

                 Thread.startVirtualThread(task);
             }
```

Thanks,
Oleh